File: typescript/src/core/xbrl/instance-document.ts
/**
 * Parser for XBRL instance documents.
 *
 * An instance document (`<base>_htm.xml` on modern filings, `<base>.xml`
 * on older ones) carries the filing's own facts: every reported value
 * with its context (period, entity, dimensions) and unit. Parsing it
 * gives as-reported data scoped to the filing, including dimensional
 * breakdowns that the aggregated company-facts API cannot provide.
 */

import { XMLParser } from "fast-xml-parser";

export interface InstanceContext {
  id: string;
  entity: string;
  startDate: string | null;
  endDate: string | null;
  instant: string | null;
  /** dimension qname -> member qname */
  dimensions: Record<string, string>;
}

export interface InstanceFact {
  concept: string;
  value: string | null;
  contextRef: string;
  unitRef: string | null;
  decimals: string | null;
}

function asArray<T>(value: T | T[] | undefined): T[] {
  if (value === undefined || value === null) {
    return [];
  }
  return Array.isArray(value) ? value : [value];
}

function localName(tag: string): string {
  return tag.includes(":") ? tag.slice(tag.indexOf(":") + 1) : tag;
}

function findKey(obj: Record<string, any>, local: string): any {
  for (const [key, value] of Object.entries(obj)) {
    if (localName(key) === local) {
      return value;
    }
  }
  return undefined;
}

function text(value: any): string {
  if (value === undefined || value === null) {
    return "";
  }
  if (typeof value === "object") {
    return String(value["#text"] ?? "").trim();
  }
  return String(value).trim();
}

export function periodKey(context: InstanceContext): string {
  return context.endDate || context.instant || "";
}

export class InstanceDocument {
  public readonly contexts: Map<string, InstanceContext>;
  public readonly units: Map<string, string>;
  public readonly facts: InstanceFact[];
  private byConceptIndex: Map<string, InstanceFact[]> | null = null;

  private constructor(
    contexts: Map<string, InstanceContext>,
    units: Map<string, string>,
    facts: InstanceFact[],
  ) {
    this.contexts = contexts;
    this.units = units;
    this.facts = facts;
  }

  static parse(raw: string): InstanceDocument {
    const parser = new XMLParser({
      ignoreAttributes: false,
      parseTagValue: false,
      trimValues: true,
    });
    const document = parser.parse(raw);

    const rootKey = Object.keys(document).find((k) => localName(k) === "xbrl");
    if (!rootKey) {
      throw new Error("Not an XBRL instance document");
    }
    const root = document[rootKey];

    const contexts = InstanceDocument.parseContexts(root);
    const units = InstanceDocument.parseUnits(root);
    const facts = InstanceDocument.parseFacts(root);
    return new InstanceDocument(contexts, units, facts);
  }

  private static parseContexts(
    root: Record<string, any>,
  ): Map<string, InstanceContext> {
    const contexts = new Map<string, InstanceContext>();
    for (const elem of asArray(findKey(root, "context"))) {
      const context: InstanceContext = {
        id: String(elem["@_id"] || ""),
        entity: "",
        startDate: null,
        endDate: null,
        instant: null,
        dimensions: {},
      };

      const entity = findKey(elem, "entity");
      if (entity) {
        context.entity = text(findKey(entity, "identifier"));
        const segment = findKey(entity, "segment");
        if (segment) {
          for (const member of asArray(findKey(segment, "explicitMember"))) {
            const dimension = String(member["@_dimension"] || "");
            const value = text(member);
            if (dimension && value) {
              context.dimensions[dimension] = value;
            }
          }
        }
      }
      const scenario = findKey(elem, "scenario");
      if (scenario) {
        for (const member of asArray(findKey(scenario, "explicitMember"))) {
          const dimension = String(member["@_dimension"] || "");
          const value = text(member);
          if (dimension && value) {
            context.dimensions[dimension] = value;
          }
        }
      }

      const period = findKey(elem, "period");
      if (period) {
        context.startDate = text(findKey(period, "startDate")) || null;
        context.endDate = text(findKey(period, "endDate")) || null;
        context.instant = text(findKey(period, "instant")) || null;
      }

      contexts.set(context.id, context);
    }
    return contexts;
  }

  private static parseUnits(root: Record<string, any>): Map<string, string> {
    const units = new Map<string, string>();
    const measureText = (value: any): string => {
      const raw = text(value);
      return raw.includes(":") ? raw.slice(raw.indexOf(":") + 1) : raw;
    };

    for (const elem of asArray(findKey(root, "unit"))) {
      const id = String(elem["@_id"] || "");
      const divide = findKey(elem, "divide");
      if (divide) {
        const numerator = measureText(
          findKey(findKey(divide, "unitNumerator") || {}, "measure"),
        );
        const denominator = measureText(
          findKey(findKey(divide, "unitDenominator") || {}, "measure"),
        );
        units.set(id, `${numerator}/${denominator}`);
      } else {
        units.set(id, measureText(findKey(elem, "measure")));
      }
    }
    return units;
  }

  private static parseFacts(root: Record<string, any>): InstanceFact[] {
    const skip = new Set(["context", "unit", "schemaRef", "footnoteLink"]);
    const facts: InstanceFact[] = [];
    const seen = new Set<string>();

    for (const [tag, value] of Object.entries(root)) {
      if (tag.startsWith("@_") || skip.has(localName(tag))) {
        continue;
      }
      for (const entry of asArray(value)) {
        if (typeof entry !== "object" || entry === null) {
          continue;
        }
        const contextRef = entry["@_contextRef"];
        if (!contextRef) {
          continue;
        }
        const unitRef = entry["@_unitRef"] ?? null;
        const key = `${tag}|${contextRef}|${unitRef}`;
        if (seen.has(key)) {
          continue;
        }
        seen.add(key);

        const nil = entry["@_xsi:nil"];
        facts.push({
          concept: tag,
          value: nil === "true" ? null : text(entry),
          contextRef: String(contextRef),
          unitRef: unitRef === null ? null : String(unitRef),
          decimals: entry["@_decimals"] ?? null,
        });
      }
    }
    return facts;
  }

  factsFor(concept: string): InstanceFact[] {
    if (this.byConceptIndex === null) {
      this.byConceptIndex = new Map();
      for (const fact of this.facts) {
        const list = this.byConceptIndex.get(fact.concept) || [];
        list.push(fact);
        this.byConceptIndex.set(fact.concept, list);
      }
    }
    return this.byConceptIndex.get(concept) || [];
  }

  concepts(): string[] {
    this.factsFor("");
    return Array.from(this.byConceptIndex!.keys());
  }

  unitOf(fact: InstanceFact): string | null {
    if (fact.unitRef === null) {
      return null;
    }
    return this.units.get(fact.unitRef) || fact.unitRef;
  }

  contextOf(fact: InstanceFact): InstanceContext | undefined {
    return this.contexts.get(fact.contextRef);
  }

  numericValue(fact: InstanceFact): number | null {
    if (fact.value === null || fact.value === "") {
      return null;
    }
    const value = Number(fact.value);
    return Number.isNaN(value) ? null : value;
  }
}

File: python/src/sec_edgar_toolkit/core/xbrl/instance_document.py
"""
Parser for XBRL instance documents.

An instance document (``<base>_htm.xml`` on modern filings, ``<base>.xml``
on older ones) carries the filing's own facts: every reported value with
its context (period, entity, dimensions) and unit. Parsing it gives
as-reported data scoped to the filing, including dimensional breakdowns
that the aggregated company-facts API cannot provide.

Standard library only. No I/O happens in this module.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Dict, List, Optional
from xml.etree import ElementTree as ET

XBRLI_NS = "http://www.xbrl.org/2003/instance"
LINK_NS = "http://www.xbrl.org/2003/linkbase"
XBRLDI_NS = "http://xbrl.org/2006/xbrldi"
XSI_NS = "http://www.w3.org/2001/XMLSchema-instance"

_XMLNS_RE = re.compile(r'xmlns:([\w.-]+)="([^"]+)"')


@dataclass
class Context:
    """One xbrli:context: entity, period, and dimensional qualifiers."""

    id: str
    entity: str = ""
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    instant: Optional[str] = None
    #: dimension qname -> member qname
    dimensions: Dict[str, str] = field(default_factory=dict)

    @property
    def is_duration(self) -> bool:
        return self.end_date is not None

    @property
    def period_key(self) -> str:
        """The period the context reports on (end date or instant)."""
        return self.end_date or self.instant or ""

    @property
    def is_dimensional(self) -> bool:
        return bool(self.dimensions)


@dataclass
class InstanceFact:
    """One reported fact from the instance document."""

    concept: str
    value: Optional[str]
    context_ref: str
    unit_ref: Optional[str] = None
    decimals: Optional[str] = None
    fact_id: Optional[str] = None

    def numeric_value(self) -> Optional[float]:
        if self.value is None:
            return None
        try:
            return float(self.value)
        except ValueError:
            return None


class InstanceDocument:
    """
    A parsed XBRL instance document.

    Attributes:
        contexts: context id -> :class:`Context`
        units: unit id -> human unit string ("USD", "shares", "USD/shares")
        facts: every reported fact, in document order
    """

    def __init__(
        self,
        contexts: Dict[str, Context],
        units: Dict[str, str],
        facts: List[InstanceFact],
    ) -> None:
        self.contexts = contexts
        self.units = units
        self.facts = facts
        self._by_concept: Optional[Dict[str, List[InstanceFact]]] = None

    @classmethod
    def parse(cls, raw: "bytes | str") -> "InstanceDocument":
        """Parse the raw XML of an instance document."""
        if isinstance(raw, str):
            raw = raw.encode("utf-8", errors="ignore")

        # ElementTree resolves tags to {uri}local; recover the document's
        # prefixes from the xmlns declarations so concepts keep their
        # familiar qnames ("us-gaap:Revenues").
        header = raw[:8192].decode("utf-8", errors="ignore")
        uri_to_prefix = {uri: prefix for prefix, uri in _XMLNS_RE.findall(header)}

        root = ET.fromstring(raw)

        contexts = cls._parse_contexts(root)
        units = cls._parse_units(root)
        facts = cls._parse_facts(root, uri_to_prefix)
        return cls(contexts, units, facts)

    # ------------------------------------------------------------------

    @staticmethod
    def _qname(tag: str, uri_to_prefix: Dict[str, str]) -> Optional[str]:
        if not tag.startswith("{"):
            return tag
        uri, _, local = tag[1:].partition("}")
        prefix = uri_to_prefix.get(uri)
        if prefix is None:
            return None
        return f"{prefix}:{local}"

    @classmethod
    def _parse_contexts(cls, root: ET.Element) -> Dict[str, Context]:
        contexts: Dict[str, Context] = {}
        for elem in root.findall(f"{{{XBRLI_NS}}}context"):
            context = Context(id=elem.get("id", ""))

            identifier = elem.find(f"{{{XBRLI_NS}}}entity/{{{XBRLI_NS}}}identifier")
            if identifier is not None and identifier.text:
                context.entity = identifier.text.strip()

            period = elem.find(f"{{{XBRLI_NS}}}period")
            if period is not None:
                start = period.find(f"{{{XBRLI_NS}}}startDate")
                end = period.find(f"{{{XBRLI_NS}}}endDate")
                instant = period.find(f"{{{XBRLI_NS}}}instant")
                if start is not None and start.text:
                    context.start_date = start.text.strip()
                if end is not None and end.text:
                    context.end_date = end.text.strip()
                if instant is not None and instant.text:
                    context.instant = instant.text.strip()

            # Dimensions live under entity/segment or under scenario
            for container in ("entity/{ns}segment", "{ns}scenario"):
                path = container.replace("{ns}", f"{{{XBRLI_NS}}}").replace(
                    "entity/", f"{{{XBRLI_NS}}}entity/"
                )
                holder = elem.find(path)
                if holder is None:
                    continue
                for member in holder.findall(f"{{{XBRLDI_NS}}}explicitMember"):
                    dimension = member.get("dimension", "")
                    if dimension and member.text:
                        context.dimensions[dimension] = member.text.strip()

            contexts[context.id] = context
        return contexts

    @classmethod
    def _parse_units(cls, root: ET.Element) -> Dict[str, str]:
        units: Dict[str, str] = {}
        for elem in root.findall(f"{{{XBRLI_NS}}}unit"):
            unit_id = elem.get("id", "")

            def measure_text(node: Optional[ET.Element]) -> str:
                if node is None or not node.text:
                    return ""
                # "iso4217:USD" -> "USD"
                return node.text.strip().split(":")[-1]

            divide = elem.find(f"{{{XBRLI_NS}}}divide")
            if divide is not None:
                numerator = divide.find(
                    f"{{{XBRLI_NS}}}unitNumerator/{{{XBRLI_NS}}}measure"
                )
                denominator = divide.find(
                    f"{{{XBRLI_NS}}}unitDenominator/{{{XBRLI_NS}}}measure"
                )
                units[unit_id] = (
                    f"{measure_text(numerator)}/{measure_text(denominator)}"
                )
            else:
                units[unit_id] = measure_text(elem.find(f"{{{XBRLI_NS}}}measure"))
        return units

    @classmethod
    def _parse_facts(
        cls, root: ET.Element, uri_to_prefix: Dict[str, str]
    ) -> List[InstanceFact]:
        skip_namespaces = (f"{{{XBRLI_NS}}}", f"{{{LINK_NS}}}")
        facts: List[InstanceFact] = []
        seen: set = set()

        for elem in root:
            tag = elem.tag
            if not isinstance(tag, str) or tag.startswith(skip_namespaces):
                continue
            context_ref = elem.get("contextRef")
            if not context_ref:
                continue

            concept = cls._qname(tag, uri_to_prefix)
            if concept is None:
                continue

            nil = elem.get(f"{{{XSI_NS}}}nil")
            value = None if nil == "true" else (elem.text or "").strip()

            unit_ref = elem.get("unitRef")
            key = (concept, context_ref, unit_ref)
            if key in seen:
                continue
            seen.add(key)

            facts.append(
                InstanceFact(
                    concept=concept,
                    value=value,
                    context_ref=context_ref,
                    unit_ref=unit_ref,
                    decimals=elem.get("decimals"),
                    fact_id=elem.get("id"),
                )
            )
        return facts

    # ------------------------------------------------------------------

    def facts_for(self, concept: str) -> List[InstanceFact]:
        """All facts reported for one concept qname."""
        if self._by_concept is None:
            index: Dict[str, List[InstanceFact]] = {}
            for fact in self.facts:
                index.setdefault(fact.concept, []).append(fact)
            self._by_concept = index
        return self._by_concept.get(concept, [])

    def concepts(self) -> List[str]:
        """Every concept qname reported in the document."""
        if self._by_concept is None:
            self.facts_for("")
        assert self._by_concept is not None
        return list(self._by_concept.keys())

    def unit_of(self, fact: InstanceFact) -> Optional[str]:
        if fact.unit_ref is None:
            return None
        return self.units.get(fact.unit_ref, fact.unit_ref)

    def context_of(self, fact: InstanceFact) -> Optional[Context]:
        return self.contexts.get(fact.context_ref)
